_drafts/cortex-m-watchpoints.md

```diff
@@ -0,0 +1,1097 @@
+
+```
+(gdb) mon reset
+(gdb) continue
+(gdb) break accel_register_sample_handler
+(gdb) break accel_example
+(gdb) c
+Continuing.
+
+Breakpoint 1, accel_register_watcher (data_processed_cb=0x469 <accel_data_processed>) at ./src/accel_fake.c:17
+17      AccelSampleProcessedCallback data_processed_cb) {
+(gdb)
+```
+
+It looks like we are registering the handler, `accel_data_processed`, as expected. Let's continue and call accel_example again and we should hit our second breakpoint:
+
+```bash
+shell> accel_example
+# debugger should have halted at
+# accel_process_reading()
+```
+
+```
+Breakpoint 2, accel_process_reading (x=x@entry=7, y=y@entry=1, z=z@entry=1) at ./src/accel_fake.c:22
+22    EXAMPLE_LOG("Processing Accel Reading ...");
+(gdb) p s_data_processed_cb
+$3 = (AccelSampleProcessedCallback) 0xffeeffee
+```
+
+We see `s_data_processed_cb` looks like a bogus address, `0xffeeffee`, but we verified it was set correctly on boot. Confusing! Let's add a watchpoint to track down when the variable is getting set:
+
+```
+(gdb) watch *(uint32_t*)&s_data_processed_cb
+(gdb) mon reset
+(gdb) c
+
+
+Hardware watchpoint 1: *(uint32_t*)&s_data_processed_cb
+
+Old value = 4293853166
+New value = 0
+0x00000a6a in prv_cinit () at ./src/startup.c:33
+33    for (uint32_t *dst = &_sbss; dst < &_ebss;) {
+```
+
+This makes sense, `s_data_processed_cb` is getting set to zero when we initialize BSS. Onward ...
+
+```
+(gdb) continue
+
+Hardware watchpoint 1: *(uint32_t*)&s_data_processed_cb
+
+Old value = 0
+New value = 1129
+0x000002e2 in accel_register_watcher (data_processed_cb=0x469 <accel_data_processed>) at ./src/accel_fake.c:18
+18    s_data_processed_cb = data_processed_cb;
+```
+
+This also makes sense, this is where we set the correct value. Onward ...
+
+```
+(gdb)
+
+Hardware watchpoint 1: *(uint32_t*)&s_data_processed_cb
+
+Old value = 1129
+New value = 65518
+0x0000040e in graphics_boot (buf=buf@entry=0x20002080 <s_graphics_buf>, len=len@entry=4) at ./src/graphics_fake.c:5
+5     for (size_t i = 0; i < len; i++) {
+```
+
+This is suspicious! The memory where we store the accel data handler is getting updated from our graphics stack. Let's examine the code
+
+```
+(gdb) list graphics_boot
+
+4	void graphics_boot(uint16_t *buf, size_t len) {
+5     for (size_t i = 0; i < len; i++) {
+6       buf[i] = 0xffee;
+7     }
+```
+
+It looks like the `len` for the `buf` argument may be incorrect.
+
+```
+(gdb) list s_graphics_buf
+[...]
+40	static uint16_t s_graphics_buf[2];
+[...]
+```
+
+Now we can see the issue! `s_graphics_buf` was defined as a `uint16_t` and we are passing the total size 4. However the function expected the array length, 2. This means we are writing past the end of the buffer into whatever variable comes after it. We can confirm in GDB that `s_registered_handler` has been allocated right after `s_graphics_buf` and is getting corrupted:
```

Comment:
Sorry I'm finding this sentence super confusing:

> Now we can see the issue! `s_graphics_buf` was defined as a `uint16_t` and we are passing the total size 4. However the function expected the array length, 2.

isn't array of 2 uint16_t the same as 4 bytes? The wording is confusing, and I'm not sure where the "4" comes from in the code...

Reply:
I've tried to rework it ... the bug I was using as an example here was a situation where ARRAY_SIZE() was expected but sizeof() was passed instead.